File: backend/internal/collection/code.go
package collection

import (
	"errors"
	"fmt"
	"strings"

	gonanoid "github.com/matoous/go-nanoid/v2"
)

const (
	codeAlphabet  = "23456789ABCDEFGHJKLMNPQRSTUVWXYZ"
	CodeLength    = 8
	codeSeparator = "-"
)

var ErrInvalidCode = errors.New("código de coleção inválido")

func NewCode() (string, error) {
	code, err := gonanoid.Generate(codeAlphabet, CodeLength)
	if err != nil {
		return "", fmt.Errorf("gerar código da coleção: %w", err)
	}

	return code, nil
}

func ParseCode(raw string) (string, error) {
	normalized := strings.ToUpper(strings.TrimSpace(raw))
	normalized = strings.ReplaceAll(normalized, codeSeparator, "")

	if len(normalized) != CodeLength {
		return "", ErrInvalidCode
	}

	for _, r := range normalized {
		if !strings.ContainsRune(codeAlphabet, r) {
			return "", ErrInvalidCode
		}
	}

	return normalized, nil
}

File: frontend/src/features/collection/code.ts
const codeAlphabet = '23456789ABCDEFGHJKLMNPQRSTUVWXYZ'
const codeSeparator = '-'

export const codeLength = 8

export function normalizeCode(raw: string): string {
  return raw.trim().toUpperCase().replaceAll(codeSeparator, '')
}

export function formatCode(raw: string): string {
  const code = normalizeCode(raw)

  if (code.length !== codeLength) {
    return code
  }

  return `${code.slice(0, codeLength / 2)}${codeSeparator}${code.slice(codeLength / 2)}`
}

export function isValidCode(raw: string): boolean {
  const code = normalizeCode(raw)

  if (code.length !== codeLength) {
    return false
  }

  return [...code].every((character) => codeAlphabet.includes(character))
}
